### PR TITLE
Attempt to convert parameter types in validator

### PR DESCRIPTION
**Pull request by replay**
We've seen many cases where users make understandable mistakes, such as specifying boolean parameters as string (`'true'`) or specifying the `inf` value as a string (`'inf'`). 

Now the input validation doesn't only check whether the input is valid, but if possible it also does a type cast of the given parameters. All the validators are now either returning a value of the correct type or raise an exception.

### CODE REVIEW

**Comment by replay** [comment]
I fixed all except 1 issue which deep source was complaining about. 
The one I didn't fix was `Exception caught is very general`:

https://deepsource.io/gh/graphite-project/graphite-web/run/daf9f1d7-3022-411d-95eb-12496f25cc22/python/PYL-W0703/

Because I think in this case we do want to catch any possible type of exception. These exceptions which we're catching there are caused by bad user input, which is hardly predictable.

**Comment by DanCech** [comment]
I suspect that there are a limited number of error types that casting to int or float can raise, `TypeError` and `ValueError` are the 2 that come to mind, I'm not sure if there are any others.

**Comment by replay** [comment]
Hmm you're right that the number of possible exceptions is probably limited. I just couldn't find any docs which clearly define all the possible exceptions raised by `int()` and `float()`. Apart from `ValueError` and `TypeError` there's also `OverflowError` which may be raised by `float()`. Also, in the methods called by `ParamTypeAggFunc.validateAggFuncs()` there are some list operations and I'm not sure what all the exceptions are which these could raise.

So basically I'm just worried I'd miss one and then we end up with a `5xx` due to bad user input, which is what we want to prevent.

**Comment by DanCech** [comment]
Since you're already catching Exception on line 232 it's probably safe enough to be more specific in the individual validation functions, since anything they miss will be handled there.

**Comment by replay** [comment]
That's a good point. I tried to come up with reasonable lists of possible exceptions, but had to change the one exception type which `getUnitString()` raises: https://github.com/graphite-project/graphite-web/pull/2574/commits/c679558f190dd870cb88d4cca0871dd249cadc50

File: webapp/graphite/functions/params.py [hunk 1]
```diff
@@ -7,54 +7,83 @@
...
+        if value.lower() == 'true':
+            return True
+        if value.lower() == 'false':
+            return False
+        raise ValueError('Invalid boolean value: {value}'.format(value=repr(value)))
+
+    if type(value) in [int, float]:
+        if value == 0:
+            return False
+        return True
```
Comment (ploxiln) [inline webapp/graphite/functions/params.py]: Maybe this should not accept all non-zero numeric values as `True`? maybe just `1` to go along with `0`?

I personally lean towards no numbers here, just `True`/`False` or `"true"`/`"false"`. While good error responses with found/expected types are definitely great, I fear that too much lenience and auto-conversion may ultimately be more confusing - things may seem to work but the results are totally wrong, or return an error that doesn't make sense, and it's because there was a mixup in a different argument which was auto-converted.

Comment (DanCech) [reply webapp/graphite/functions/params.py]: Good point inre being too permissive, I think only 0/1 is a good compromise here.

File: webapp/graphite/functions/params.py [hunk 2]
```diff
@@ -40,7 +40,9 @@ def validateBoolean(value):
     if type(value) in [int, float]:
         if value == 0:
             return False
-        return True
+        if value == 1:
+            return True
+        raise ValueError('Invalid numeric value: {value}'.format(value=repr(value)))
```
Comment (ploxiln) [inline webapp/graphite/functions/params.py]: should be "Invalid boolean value"?

Comment (replay) [reply webapp/graphite/functions/params.py]: right, updated

File: webapp/graphite/functions/params.py (hunk 1 above)
Comment (replay) [reply webapp/graphite/functions/params.py]: sounds good 04ec05f6b6c37dc520e59316c818890666b99e9f

Comment (ploxiln) [reply webapp/graphite/functions/params.py]: looks good, thanks :)

**Comment by replay** [comment]
@deniszh If you have a moment it would be great if you could take a look at this PR

**Comment by deniszh** [comment]
Ah, sorry, I thought I reviewed it.
